apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/EphemeralTransaction.cs

```diff
@@ -127,5 +119,45 @@ async Task<ReliableStreamReader> Impl()
                 return reader;
             }
         }
+
+        /// <summary>
+        /// Creates a read/write transaction and executes a task within it.
+        /// The transaction is rolled back if the action provided when creating the task has not been called.
+        /// </summary>
+        private Task<T> ExecuteInReadWriteTransactionWithErrorTranslationAndProfiling<T>(Func<SpannerTransaction, Action, Task<T>> taskProvider, CancellationToken cancellationToken, string name)
+        {
+            return ExecuteHelper.WithErrorTranslationAndProfiling(Impl, name, _connection.Logger);
+
+            async Task<T> Impl()
+            {
+                using (var transaction = await _connection.BeginTransactionImplAsync(_transactionOptions, TransactionMode.ReadWrite, cancellationToken).ConfigureAwait(false))
+                {
+                    bool rollback = true;
+                    try
+                    {
+                        return await taskProvider(transaction, () => rollback = false).ConfigureAwait(false);
+                    }
+                    finally
+                    {
+                        if (rollback)
```

Comment:
Why not simply catch any exception and _attempt_ a rollback? Is it because of partioned-dml? What happens if we rollback that transaction? I'd say it would rollback to the last commit which is fine?

Reply:
It's because I believe we shouldn't call Rollback if we've already called Commit - that a failed Commit already rolls back.
But let's wait to hear more.